core/src/main/java/org/infinispan/util/concurrent/locks/impl/DefaultLockManager.java

```diff
@@ -0,0 +1,378 @@
+         }
+      }
+      compositeLockPromise.markListAsFinal();
+      scheduleLockPromise(compositeLockPromise, time, unit);
+      return compositeLockPromise;
+   }
+
+   private Set<Object> filterDistinctKeys(Collection<?> collection) {
+      if (collection instanceof Set) {
+         //noinspection unchecked
+         return (Set<Object>) collection;
+      } else {
+         return new HashSet<>(collection);
+      }
+   }
+
+   @Override
+   public void unlock(Object key, Object lockOwner) {
+      if (trace) {
+         log.tracef("Release lock for key=%s. owner=%s", key, lockOwner);
+      }
+      lockContainer.release(key, lockOwner);
+   }
+
+   @Override
+   public void unlockAll(Collection<?> keys, Object lockOwner) {
+      if (trace) {
+         log.tracef("Release locks for keys=%s. owner=%s", keys, lockOwner);
+      }
+      if (keys.isEmpty()) {
+         return;
+      }
+      for (Object key : keys) {
+         lockContainer.release(key, lockOwner);
+      }
+   }
+
+   @Override
+   public void unlockAll(InvocationContext context) {
+      unlockAll(context.getLockedKeys(), context.getLockOwner());
+      context.clearLockedKeys();
+   }
+
+   @Override
+   public boolean ownsLock(Object key, Object lockOwner) {
+      Object currentOwner = getOwner(key);
+      return currentOwner != null && currentOwner.equals(lockOwner);
+   }
+
+   @Override
+   public boolean isLocked(Object key) {
+      return getOwner(key) != null;
+   }
+
+   @Override
+   public Object getOwner(Object key) {
+      InfinispanLock lock = lockContainer.getLock(key);
+      return lock == null ? null : lock.getLockOwner();
+   }
+
+   @Override
+   public String printLockInfo() {
+      return lockContainer.toString();
+   }
+
+   @Override
+   @ManagedAttribute(description = "The number of exclusive locks that are held.", displayName = "Number of locks held")
+   public int getNumberOfLocksHeld() {
+      return lockContainer.getNumLocksHeld();
+   }
+
+   @ManagedAttribute(description = "The concurrency level that the MVCC Lock Manager has been configured with.", displayName = "Concurrency level", dataType = DataType.TRAIT)
+   public int getConcurrencyLevel() {
+      return configuration.locking().concurrencyLevel();
+   }
+
+   @ManagedAttribute(description = "The number of exclusive locks that are available.", displayName = "Number of locks available")
+   public int getNumberOfLocksAvailable() {
+      return lockContainer.size() - lockContainer.getNumLocksHeld();
+   }
+
+   @Override
+   public InfinispanLock getLock(Object key) {
+      return lockContainer.getLock(key);
+   }
+
+   private void scheduleLockPromise(LockPromise promise, long time, TimeUnit unit) {
+      if (!promise.isAvailable() && time > 0 && scheduler != null) {
+         final ScheduledFuture<?> future = scheduler.schedule(promise::isAvailable, time, unit);
```

Comment:
Also, see how I've avoided allocating new lambdas for the timeout task and cancellation here: https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/remoting/transport/jgroups/SingleResponseFuture.java